smfile_writer.py

```diff
@@ -1,152 +1,152 @@
-            n = pow
-
-    k = [0]*int(256/(2**n))
-    for p in position:
-        f = int(p/(2**n))-1
-        k[f] = 1
-
-    return k
-
-def calculate_measure(measure, timing):
-    trim    = []
-
-    if not measure:
-        trim = [0,0,0,0,',']
-    else:
-        trim = [0] * 256
-        for note in measure:
-            n   = 0
-            min = 10.0
-            for i in range(len(trim)):
-                diff = abs(note - i*timing)
-                if diff <= min:
-                    n = i
-                    min = diff
-            trim[n] = 1
-        trim = trim_measure(trim)
-        trim.append(',')
-    return trim
-
-def calculate_notes(timings, bpm):
-    note = []
-    if not timings:
-        return note
-    measure_sec     = round(4 * 60/bpm, 10)                                                       #measure in seconds = 4 x seconds per beat
-    measure_all     = math.ceil(timings[-1]/measure_sec)                                          #number of measures that fits the whole song
-    note_256        = round(measure_sec/256, 5)                                                    #time of notes in seconds
-
-    for i in range(measure_all):
-        measure = []
-        for j in timings:
-            if((i * measure_sec) <= j < ((i+1) * measure_sec)):
-                measure.append(round(j-(i*measure_sec), 5))
-        note.extend(calculate_measure(measure, note_256))
-    note[-1] = ';'
-    return note
+    # pre-generate output data
+    title  = '#TITLE:%s;\n' % step_dict['title']
+    artist = '#ARTIST:jhaco vs cpuguy96;\n'
+    music  = '#MUSIC:%s.ogg;\n' % file_name
+    select = 'SELECTABLE:YES;\n'
+    bpm    = 'BPMS:0.000=%s;\n\n' % str(step_dict['bpm'])
+    note_data = ''
+    for difficulty in step_dict['notes'].keys():
+        note_data += '//---------------dance-single - ----------------\n'
+        note_data += '#NOTES:\n'
+        note_data += '     dance-single:\n'
+        note_data += '     :\n'
+        note_data += '     %s:\n' % difficulty
+        note_data += '     8:\n'
+        note_data += '     1.000,1.000,1.000,1.000,1.000:\n'
+        for note in step_dict['notes'][difficulty]:
+                note_data += note + '\n'
+
+    with open(join(output_dir, ofile), 'w') as f:
+        f.write(''.join((title, artist, music, select, bpm, note_data))) # calling write only once is best practice
+        
+#===================================================================================================
+
+def find_gcd(note_positions):
+    x = reduce(gcd, note_positions + [256])
+    return x
+
+def generate_measure(notes, note_positions):
+    # tries to fit 256, 128, 64, 32, 16, 8 or 4 note measures
+    note_gcd = find_gcd(note_positions)
+
+    if note_gcd == 128: # if fitting returns a 2 note measure, adjust to 4 note
+        note_gcd = 64
+
+    # place notes in generated measure
+    generated_measure = ['0000']*int(256/note_gcd)
+    for note, p in enumerate(note_positions):

```

Comment:
This can be rewritten to:
`for note, note_position in zip(notes, note_positions)`